support new snapd_refresh from layer-snap

kwmonroe opened the pull request:
Default our etcd charm to use the 'max' possible refresh window,
which is currently to refresh once a month. The leader will get
the initial value and set leader data so all followers (present
and future) will refresh around the same time.

Adding an amulet test for this and fixed up lint issues in the tests.

kwmonroe committed:
support new snapd_refresh from layer-snap

Default our etcd charm to use the 'max' possible refresh window,
which is currently to refresh once a month. The leader will get
the initial value and set leader data so all followers (present
and future) will refresh around the same time.

Adding an amulet test for this and fixed up lint issues in the tests.

kwmonroe committed:
snap.get works now, use it

hyperbolic2346 commented on reactive/etcd.py:
```diff
@@ -414,6 +415,35 @@ def initialize_new_leader():
+@when('leadership.changed.snapd_refresh')
+@when_not('leadership.is_leader')
+def set_snapd_timer():
+    ''' Set the snapd refresh.timer on non-leader cluster members. '''
```
Do we need to handle where I'm not the leader, but the leader hasn't set snapd_refresh yet?

kwmonroe replied:
I think we're covered with `@when('leadership.changed.snapd_refresh')`.  If the leader hasn't set it, the leadership.changed flag won't be set.

If you're talking about the case where a follower joins and doesn't know about the leader yet, then yes, layer-snap will set the refresh to a month from yesterday.  But that will quickly be re-set (if it even has a chance to *be* set) as soon as the follower handles the leader hooks.

So, no, i don't think we need a handler for `@when_not(leader)` and `@when_not(leadership.data)`.

kwmonroe committed:
gate snapd_refresh handlers on snap.refresh.set

If snapd_refresh handlers from layer-etcd run before layer-snap,
followers may set the refresh config to the leader's data and
subsequently have it overwritten by layer-snap.

Prevent this by gating our refresh handlers with 'snap.refresh.set':

https://git.launchpad.net/layer-snap/commit/?h=defer-refresh&id=62d918f8979451128af93973aa57635e19287fee

This ensures layer-snap is done configuring the core snap before we
process our leader config. This is also useful on the leader so we
don't bother trying to get the config until we know it's set.
